GraphQL endpoint

VarshiniSampath opened the pull request:
@michael-merchant @cabloo

Commit:
GraphQL endpoint

VarshiniSampath commented:
I've exported the MongoDB database. So you can test it by importing the data/papers.json collections in a database called 'publications'.

Use these CURL requests to try out the queries:

**For one paper:**
`curl -X POST -H "Content-Type: application/json" -d '{query": "{ publication { title } }"}' http://localhost:4000/graphql`

**For a list of papers:**
`curl -X POST -H "Content-Type: application/json" -d '{query": "{ publications { title } }"}' http://localhost:4000/graphql`

**For a paper given an DOI**
`curl -X POST -H "Content-Type: application/json" -d '{query": "{ publicationByDOI(doi: \"arXiv:1611.08408\") { title } }"}' http://localhost:4000/graphql`

mike-merchant commented:
Can you add notes in the PR description about how you were able to test this? e.g. curl commands, graphiql, etc.

mike-merchant commented on data/models/mongooseSchema.js:
```diff
@@ -0,0 +1,19 @@
+import mongoose from 'mongoose';
+
+
```
nit: remove extra newline

mike-merchant commented on data/schema.js:
```diff
@@ -0,0 +1,23 @@
+import { buildSchema } from 'graphql';
+import { GraphQLList } from 'graphql';
```
Since these are both from same source I believe you can comma delimit instead of putting on separate lines.

mike-merchant commented on data/schema.js:
```diff
@@ -0,0 +1,23 @@
+    title: String,
+    author: String,
+    year: Int,
+    citationCount: Int
```
nit: does trailing slash work here?

VarshiniSampath replied:
Can you explain this further?

mike-merchant replied:
Sorry, I meant trailing comma.

mike-merchant commented on data/types/publicationType.js:
```diff
@@ -0,0 +1,32 @@
+    citationCount : {
+      type: GraphQLInt,
+      description: 'Title of the paper',
+    }
```
nit: trailing slash

mike-merchant replied:
nit: trailing comma

mike-merchant commented on data/types/publicationType.js:
```diff
@@ -0,0 +1,32 @@
+    },
+    citationCount : {
+      type: GraphQLInt,
+      description: 'Title of the paper',
```
Copy-paste error w/ description

mike-merchant commented on data/types/publicationType.js:
```diff
@@ -0,0 +1,32 @@
+    },
+    year: {
+      type: GraphQLInt,
+      description: 'Title of the paper',
```
Copy-paste error with description

mike-merchant commented on data/types/publicationType.js:
```diff
@@ -0,0 +1,32 @@
+  fields: () => ({
+    doi: {
+      type: GraphQLString,
+      description: 'DOI of the paper',
```
Try not to abbreviate in descriptions

mike-merchant commented on webpack.config.js:
```diff
@@ -0,0 +1,11 @@
+  module: {
+    loaders: [{
+      exclude: /node_modules/,
+      loader: 'babel-loader'
```
Supports trailing commas?

mike-merchant reviewed (changes requested):
Just a bunch of nits. Nice work!

Commit:
Mutation added

VarshiniSampath commented:
@michael-merchant : Fixed previous review changes and added mutation endpoint with Graphql. Will push relay and react code after this gets reviewed.

**CURL command to add a new publication to DB:**
`curl -X POST http://localhost:4000/graphql -H "Contenttype: application/graphql" -d "mutation { insertPublication(input: {doi: \"arxiv:123\", author: \"John Doe\", title: \"Sample paper\", year: 2002, citationCount: 20}){title}}"`

VarshiniSampath commented:
@wesley-moy: Please take a look.

mike-merchant commented:
Other than trailing comma nits, this LGTM

cabloo commented on data/models/mongooseResolvers.js:
```diff
@@ -0,0 +1,61 @@
+const getPublication = () => {
+  console.log('Getting one publication from DB');
+  return new Promise((resolve, reject) => {
+    mongoPublicationModel.findOne({}, (err, record) => {
```
Does `findOne` return a promise if not passed a callback? If so `new Promise(...)` would be unnecessary and this file would look a lot cleaner.

VarshiniSampath replied:
findOne doesn't return a promise directly. We should either use it with exec() or new Promise().

cabloo commented on data/models/mongooseResolvers.js:
```diff
@@ -0,0 +1,61 @@
+// This file has resolvers for queries that get data from MongoDB using Mongoose.
```
I would make this file a `Publication` class with static methods `getOne`, `list`, `getByDOI`, etc.

cabloo commented on data/schema.js:
```diff
@@ -0,0 +1,34 @@
+import { buildSchema, GraphQLList } from 'graphql';
```
I think that in my app this file was generated from other files - not sure if that is the case here but if so it should not be included in source control.

VarshiniSampath replied:
schema.js is our own file, it's not generated. schema.graphql is the generated file and the way we handle that is different. You've generated it through the updateSchema.js script. I, on the other hand, have generated it with the buildSchema method in 'graphql' module.

cabloo reviewed (commented):
minor comments
